### PR TITLE
Issue #6400 - Add config option to enable/disable dynamic plugin reload feature 

### PR DESCRIPTION
**Pull request by 15ljindal**
Adds config option "proxy.config.plugin.dynamic_reload_mode" which can take values 0 (disable) or 1 (enable). The default value will be 0.
This will be applicable only for remap plugins (global plugins do not yet support dynamic reload).
The assumption is that if dynamic reload is disabled, it will be disabled for all plugins specified in remap.config.
This is applicable for ATS 9 (branch 9.0.x)

### CODE REVIEW

**Comment by randall** [comment]
[approve ci]

**Comment by vmamidi** [comment]
This can't be overloaded?

**Comment by 15ljindal** [comment]
> This can't be overridden?

Sorry, but I am not sure what exactly you are asking here. We have this config parameter which will simply turn off this feature so that we have old-style loading of all plugins mentioned in remap.config.

**Comment by vmamidi** [comment]
> > This can't be overridden?
> 
> Sorry, but I am not sure what exactly you are asking here. We have this config parameter which will simply turn off this feature so that we have old-style loading of all plugins mentioned in remap.config.

Can it be changed to turn on or off per remap rule?

**Comment by gtenev** [comment]
@15ljindal, thank you for working on this PR! Here is the requested review.

To achieve plugin isolation ATSv9 opens with `dlopen()`a temporary DSO file copy. Hence adding an on-off-switch to skip the copying is obviously the right direction. 

I wonder if you would agree to change your implementation after considering the following points.

####  Reduce the scope of the option?

If we implement the on-off-switch in `PluginDso` we prevent practically the isolation/reloading of all plugin types regardless. What we actually need here is a on-off-switch only for global and remap plugins (for compatibility).

ATSv9 plugin loading infrastructure that enabled the “plugin reload” feature was also meant to unify all DSO loading mechanisms in ATS. One of the design goals was to allow eventually to move global plugins over and use the same infrastructure to (re)load other non-global and non-remap plugins types, one example could be @jrushford's parent selection plugins.

#### Rename configuration option?

Instead of thinking in terms of switching on and off the “plugin reload” maybe we could  think more in terms of switching on and off the isolation between the plugins of different types and versions. The isolation is just prerequisite for "plugin reload", not the same thing.

May be `disable_global_remap_plugin_isolation` or something like that? 
Better names welcome? (naming is one of the 3 challenges in computer science).


#### isPluginDynamicReloadEnabled out of PluginDso?

My personal preference would be to move  the functionality of `isPluginDynamicReloadEnabled()` out of `PluginDso` because it introduces unnecessary dependency in PluginDso on the ATS configuration infrastructure (i.e. think about writing a unit test for the new option, future refactoring, etc).


#### Design / implementation idea

There is probably no single best way but may be we could do something like the following.

```
# In plugin factory user code (caller)
bool plugin_reload_enabled =  CallerClassOrNameSpace::isPluginDynamicReloadEnabled();
RemapPluginInst *pi = PluginFactory::getRemapPlugin(…, plugin_reload_enabled); 
```

`PluginFactory::getRemapPlugin()` based on `plugin_reload_enabled` could then set the plugin effective and runtime paths accordingly (same or different) 

`PluginDso::load()` could skip copying of the DSO file if plugin effective and runtime paths are the same (which would disable practically the isolation bringing back the old behavior).


In such a way the plugin loading design / implementation would not prevent us to
- isolate / reload other types of plugins.
- go with any type of option - “compile-time”, “run-time”, “overridable” / “non-overridable”, on or off by default (whatever we decide we want to do here)


#### Unit tests and documentation

Please make sure the existing "plugin reload" behavior still works / unit tests still pass and also add unit tests testing the isolation effect of the new option. The new plugin loading infrastructure has a pretty decent coverage (IMHO), let's keep it that way! 

Thank you for adding config option documentation! It will be great to add a paragraph about the isolation problem between global and remap plugins in `reloading-plugins.en.html` design document (something that we missed to explained initially).

**Comment by sudheerv** [comment]
@gtenev @vmamidi - We need something quick to unblock us that will preserve the old behavior. We have no plans of using dynamic plugin reload in the near future and Lokesh’s current patch allows us to unblock our ATS9 migration effort and move us to have it in production sooner where I’m sure we will see other issues that will need troubleshooting. 

I agree with Vijay’s suggestion on making this overridable but given the time crunch we are in, this probably is a decent start. We can always extend it further in the future. Unless there are any fundamental issues you see with the patch, would be great if you can review it from the perspective of KISS at this point. We have already applied this in our internal ats9 and are moving forward with this patch at this point. 

@15ljindal pls feel free to correct me if I mis-stated anything :)

File: proxy/Plugin.h [hunk 1]
```diff
@@ -26,6 +26,15 @@
 #include <string>
 #include "tscore/List.h"
 
+enum plugin_dynamic_reload_mode {
```
Comment (SolidWallOfCode) [inline proxy/Plugin.h]: This is C++, there is no need to `typedef` or `using` an `enum`. Just name it "PluginDynamicReloadMode".

File: proxy/http/remap/PluginDso.cc [hunk 2]
```diff
@@ -332,3 +355,17 @@ PluginDso::LoadedPlugins::indicatePostReload(bool reloadSuccessful, const std::u
 }
 
 Ptr<PluginDso::LoadedPlugins> PluginDso::_plugins;
+
+bool
+isPluginDynamicReloadEnabled()
```
Comment (SolidWallOfCode) [inline proxy/http/remap/PluginDso.cc]: I'm a bit concerned about this - how often is this method called? Might it not be better to have a global variable with the value from the configuration, which is updated when "records.config" is reloaded? You could look at `Enable_Config_Var`, `HttpEstablishStaticConfigByte`,  `http_server_session_sharing_cb`, or `http_config_enum_read`. The server session sharing configuration, or look at how HttpConnectionCount values are set up.

Comment (SolidWallOfCode) [reply proxy/http/remap/PluginDso.cc]: Looking further in the code and checking the documentation, this does not appear to be reloadable itself and is therefore an event better candidate for reading once and storing in a global variable.

File: proxy/http/remap/unit-tests/plugin_testing_common.h [hunk 3]
```diff
@@ -38,6 +38,8 @@
 
 #include "../PluginFactory.h"
 
+const bool DYNAMIC_PLUGIN_RELOAD_IS_ENABLED = true;
```
Comment (SolidWallOfCode) [inline proxy/http/remap/unit-tests/plugin_testing_common.h]: Might as well go all the way to `constexpr`.

File: proxy/http/remap/unit-tests/test_PluginFactory.cc [hunk 4]
```diff
@@ -114,17 +115,83 @@ clean()
...
+    return _dlh;
+  }
+
+  int
+  getPluginVersion()
+  {
+    std::string error;
+    void *s = nullptr;
+    CHECK(getSymbol("pluginDsoVersionTest", s, error));
+    int (*version)() = reinterpret_cast<int (*)()>(s);
```
Comment (SolidWallOfCode) [inline proxy/http/remap/unit-tests/test_PluginFactory.cc]: This is a great place for `auto`.

**Comment by SolidWallOfCode** [comment]
Definitely impressed by the testing code. I need to spend some more time reviewing that.

**Comment by gtenev** [comment]
@sudheerv @15ljindal although I used a long explanation to show where I am coming from, I don't believe the idea I gave you about an alternative implementation is any more complicated or more time consuming than what you proposed with this PR. 

What I proposed is of the same complexity and is just less limiting (reduces the scope of the option impact just to global/remap plugins). It will also allow you to do what @vmamidi was asking for - "overridable" runtime option and still allow us to disable to "plugin reload" feature for you (whichever we choose).

IMHO we need to be a little bit more forward looking than this.

File: proxy/http/remap/PluginDso.cc (hunk 2 above)
Comment (gtenev) [reply proxy/http/remap/PluginDso.cc]: Why don't we move this out of the `PluginDso` and call it once or as many times we need and pass a parameter to `PluginDso`. I think this will improve the modularity (easier to test as well, w/o the dependency on the the configuration code)

Comment (15ljindal) [reply proxy/http/remap/PluginDso.cc]: Thanks for the response Alan. I will incorporate this and other feedback from you in my updated PR. Having a global variable will work in this case. Can you suggest the best place to define and initialize he global variable?

**Comment by 15ljindal** [comment]
> Please fix/update the unit-tests and also see my comments on the PR about reducing the scope, alternative implementations.

Thanks a lot for the feedback @gtenev . 

**Scope**: The scope of this change is to make sure we achieve the legacy behavior of plugin loading and keep it as simple as possible (as @sudheerv mentioned). From the user point of view, the behavior would be - "Once user has started the traffic server process, user would not want to do any kind re-loading of a plugin. Once a plugin has been loaded (be it part of plugin.config or remap.config), user would not want to reload the same plugin Dso again." This is what the behavior was prior to "dynamic reload" feature and having that behavior is the goal of this PR.
If we want to do anything which adds more capabilities (e.g. per remap line option to enable/disable the dynamic reload) then they can be par of future work as @sudheerv mentioned.

**Option name**: In line with the above mentioned goal I feel names like "proxy.config.plugin.dynamic_reload_mode" or "proxy.config.plugin.dynamic_reload_enabled" are simpler and thus more suitable. This option would be non-reloadable and I am leaning towards using a global variable as @SolidWallOfCode suggested in the implementation. By default, the "dynamic reload" feature would be ENABLED.

**Implementation**: I see your point. As you suggested, one of the ideas I considered (and implemented first) was to just pass the bool "dynamicReloadEnabled" as part of function calls (like `PluginFactory::getRemapPlugin(..., bool dynamicReloadEnabled)` or `RemapPluginInfo::load(..., bool dynamicReloadEnabled)` or wherever needed. However, that does not scale well.

One of the most critical requirements to restore legacy behavior is that we do not end up deleting any user specified Dso file. Function `PluginDso::clean(...)` which deletes the Dso file should effectively be a NOP when dynamic reload is disabled. This function can be called on multiple code paths (from `PluginDso::unload(...)` or `PluginDso::~PluginDso()` etc. As such the most suitable way to make `PluginDso::clean(...)` a NOP is to either have a class member `bool _dynamicReloadEnabled` as in the current version of PR, or use a global variable which I can create. Same can be said about function `PluginFactory::clean(...)`

**Testing**: Already existing tests in the folder unit-tests are very comprehensive. I have made sure all those tests pass with this PR and they are practically untouched - the only change is passing the variable "bool dynamicReloadEnabled" as TRUE. So I am not exactly sure what your concern is regarding existing tests. Please let me know if I mis-read something.
Additionally, I have added some tests using the existing infra to test the legacy behavior(s) this PR is targeting.

**Documentation**: I am going to add more documentation to ` reloading-plugins.en.html` to explain this change in more detail. Thanks for pointing that out.

Please let me know your thoughts so that I can work on updating this PR and get this change in at the earliest.

**Comment by 15ljindal** [comment]
> Please fix/update the unit-tests and also see my comments on the PR about reducing the scope, alternative implementations.

I have updated the pull request based on our discussions here. Please review.

File: proxy/Plugin.h (hunk 1 above)
Comment (15ljindal) [reply proxy/Plugin.h]: Fixed this.

File: proxy/http/remap/PluginDso.cc (hunk 2 above)
Comment (15ljindal) [reply proxy/http/remap/PluginDso.cc]: fixed this.

**Comment by randall** [comment]
[approve ci]

File: proxy/http/remap/unit-tests/plugin_testing_common.h (hunk 3 above)
Comment (15ljindal) [reply proxy/http/remap/unit-tests/plugin_testing_common.h]: fixed.

File: proxy/http/remap/unit-tests/test_PluginFactory.cc [hunk 5]
```diff
@@ -114,17 +115,83 @@ clean()
   fs::remove(sandboxDir, ec);
 }
 
+// this is a simple class to simulate loading of Plugin Dso as done during loading of global plugins
+class GlobalPluginInfo
```
[the remark this answers is not in the document]
Comment (15ljindal) [reply proxy/http/remap/unit-tests/test_PluginFactory.cc]: The idea is to keep it obvious that this class mimics loading of global plugin and is thus completely independent of all the infrastructure which is part of dynamic plugin (re)loading / remap plugins. Current implementation is simple and clear, and I would prefer to keep it that way.

File: proxy/http/remap/PluginDso.cc [hunk 6]
```diff
@@ -215,6 +216,18 @@ PluginDso::modTime() const
...
 }
 
+/**
+ * @brief file handle returned by dlopen syscall
+ *
+ * @return dlopen filehandle
+ */
+
+void *
+PluginDso::dlOpenHandle() const
```
[the remark this answers is not in the document]
Comment (15ljindal) [reply proxy/http/remap/PluginDso.cc]: Yeah, I had the same thing in mind. First, I feel it is a reasonable assumption to assume that the dso will be opened using dlopen and thus asking for its handle seems reasonable as well.

I looked at how we can use the suggestion of having a class derived from PluginDso where this member can be acessed in unit-tests, and there does not seem to be an easy way to do that without writing too much bolier-plate code for a small change. I will try to explain here.
There are calls like `RemapPluginInst *plugin_v2      = factory2->getRemapPlugin(configName, 0, nullptr, error2, isPluginDynamicReloadEnabled());` which return an object of type `RemapPluginInst` which holds a pointer to an object of type `RemapPluginInfo`. In order to get get access to `_dlh` without this `PluginDso::dlOpenHandle()` function, I think I would have to define `class RemapPluginInfoTest : public RemapPluginInfo` and then define copy constructors in class `RemapPluginInfoTest`, `RemapPluginInfo` and `PluginDso`, so that I can do something like
`RemapPluginInst *plugin_v2      = factory2->getRemapPlugin(configName, 0, nullptr, error2, isPluginDynamicReloadEnabled());`
`RemapPluginInfoTest test_object(plugin_v2->_plugin);`
`CHECK(test_object._dlh == ...)`

In short, having this function `PluginDso::dlOpenHandle()` seems fine to me. Please let me know if I missed something or if you have any further thoughts.

File: proxy/http/remap/unit-tests/test_PluginFactory.cc [hunk 7]
```diff
@@ -204,32 +271,59 @@ SCENARIO("loading plugins", "[plugin][core]")
...
       THEN("expect it to successfully load")
       {
         validateSuccessfulConfigPathTest(plugin, error, effectivePath, runtimePath);
-        CHECK(nullptr != PluginDso::loadedPlugins()->findByEffectivePath(effectivePath));
+        CHECK(nullptr != PluginDso::loadedPlugins()->findByEffectivePath(effectivePath, DYNAMIC_PLUGIN_RELOAD_IS_ENABLED));
+
+        // check Dso still exists
+        CHECK(plugin->_plugin.effectivePath() == plugin->_plugin.runtimePath());
+        CHECK(fs::exists(plugin->_plugin.effectivePath()));
+        CHECK(fs::exists(plugin->_plugin.runtimePath()));
```
[the remark this answers is not in the document]
Comment (15ljindal) [reply proxy/http/remap/unit-tests/test_PluginFactory.cc]: fixed.

File: proxy/http/remap/PluginDso.h [hunk 8]
```diff
@@ -69,6 +71,7 @@ class PluginDso : public PluginThreadContext
   const fs::path &effectivePath() const;
   const fs::path &runtimePath() const;
   time_t modTime() const;
+  void *dlOpenHandle() const;
```
[the remark this answers is not in the document]
Comment (15ljindal) [reply proxy/http/remap/PluginDso.h]: please see my comment above for the same question.

File: proxy/http/remap/PluginDso.cc [hunk 9]
```diff
@@ -71,9 +71,9 @@ PluginDso::load(std::string &error)
   } else {
     PluginDebug(_tag, "plugin '%s' effective path: %s", _configPath.c_str(), _effectivePath.c_str());
 
-    /* Copy the installed plugin DSO to a runtime directory */
+    /* Copy the installed plugin DSO to a runtime directory if dynamic reload enabled */
     std::error_code ec;
-    if (!copy(_effectivePath, _runtimePath, ec)) {
+    if (isPluginDynamicReloadEnabled() && !copy(_effectivePath, _runtimePath, ec)) {
```
[the remark this answers is not in the document]
Comment (15ljindal) [reply proxy/http/remap/PluginDso.cc]: fixed.

File: proxy/Plugin.cc [hunk 10]
```diff
@@ -245,7 +266,8 @@ plugin_init(bool validateOnly)
   if (INIT_ONCE) {
     api_init();
     plugin_dir = ats_stringdup(RecConfigReadPluginDir());
-    INIT_ONCE  = false;
+    setupPluginDynamicReloadConfig();
```
[the remark this answers is not in the document]
Comment (15ljindal) [reply proxy/Plugin.cc]: fixed.

File: proxy/http/remap/unit-tests/test_PluginFactory.cc [hunk 11]
```diff
@@ -568,6 +676,128 @@ SCENARIO("loading multiple version of the same plugin at the same time", "[plugi
...
+        CHECK(fs::exists(plugin_v1->_plugin.effectivePath()));
+        CHECK(!fs::exists(plugin_v1->_plugin.runtimePath()));
+      }
+
+      teardownConfigPathTest(factory1);
+      teardownConfigPathTest(factory2);
+    }
+  }
+
+  GIVEN("two different versions v1 and v2 of same plugin - dynamic plugin reload DISABLED")
```
[the remark this answers is not in the document]
Comment (15ljindal) [reply proxy/http/remap/unit-tests/test_PluginFactory.cc]: fixed.

File: proxy/http/remap/PluginDso.cc (hunk 6 above)
[the remark this answers is not in the document]
Comment (15ljindal) [reply proxy/http/remap/PluginDso.cc]: not sure why my response does not show up here. Please see - https://github.com/apache/trafficserver/pull/6421#discussion_r383079718

File: proxy/http/remap/PluginDso.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (15ljindal) [reply proxy/http/remap/PluginDso.h]: again. please see - https://github.com/apache/trafficserver/pull/6421#discussion_r383079718

**Comment by randall** [comment]
[approve ci]

File: proxy/http/remap/PluginDso.cc (hunk 6 above)
[the remark this answers is not in the document]
Comment (gtenev) [reply proxy/http/remap/PluginDso.cc]: > First, I feel it is a reasonable assumption to assume that the dso will be opened using dlopen and thus asking for its handle seems reasonable as well.

Encapsulation can be an obstacle to simplicity but it has its benefits - think in terms of preventing the `PluginDso` user from calling `dlclose` directly, also about possibility to implement internal `PluginDso` housekeeping 

> In short, above seems like an over-kill 

I think it could be expected to have little bit more "cumbersome" unit-test code in the name of better module design. IMHO it does not mean that this is not worth pursuing, it rather means that we have not come up with a better way to do it.

> Clean and simple

True, access modifiers, design patterns often go in the opposite direction to simplicity (similarly security vs usability) but it is believed that they could help long term. Also sometimes it just seem more complex than it actually is (i.e. removing the coupling in `PluginDso` did not really make the code more complex)

> I would prefer to keep it that way.

Fair enough, I appreciate the thorough unit-tests related to the plugin isolation! I think we are in a good shape here - coverage is good and we can catch regressions in future which is the final goal here.

File: proxy/http/remap/unit-tests/test_PluginFactory.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (gtenev) [reply proxy/http/remap/unit-tests/test_PluginFactory.cc]: please see my other comment below

**Comment by sudheerv** [comment]
[approve ci autest]

**Comment by randall** [comment]
[approve ci]

**Comment by randall** [comment]
[approve ci clang-analyzer]

**Comment by sudheerv** [comment]
@15ljindal Can you take a look at the clang failure?

**Comment by randall** [comment]
[approve ci clang-analyzer]

**Comment by randall** [comment]
clang-analyzer issue was fixed by #6470.

**Comment by zwoop** [comment]
Tested on docs.
Cherry-picked to v9.0.x branch.
